src/libponyc/pass/refer.c

```diff
@@ -926,6 +926,50 @@ static bool refer_seq(pass_opt_t* opt, ast_t* ast)
   return true;
 }
 
+static bool valid_is_comparand(pass_opt_t* opt, ast_t* ast)
+{
+  ast_t* type;
+  switch(ast_id(ast))
+  {
+    case TK_TYPEREF:
+      type = ast_data(ast);
+      if(ast_id(type) != TK_PRIMITIVE)
+      {
+        ast_error(opt->check.errors, ast, "comparing a non-primitive type constructor will always fail");
```

Comment:
This is a good error message except I think the folks who are most likely to trigger it might not realize they are doing "as" against a constructor even when they see the error.

Reply:
Good point, changed it. Hopefully it's clearer now.